content/blog/2019-10-21-threads-not-as-libraries/index.md

```diff
@@ -0,0 +1,272 @@
+Intuitively, we may expect an implementation akin to the **sequential consistency** (SC) model, which interprets a threaded program as an interleaving of instructions across threads, such that intra-thread instruction order is preserved.
+
+For example, under the SC model, we would observe a data race whereby at least one of `r1 == 1` or `r2 == 1` holds after the execution of the following threaded program:
+
+
+```c
+int r1 = 0;
+int r2 = 0;
+
+// Thread 1
+x = 1; r1 = y;
+
+// Thread 2
+y = 1; r1 = x;
+```
+
+SC is an example of a **memory model**, a specification for the behavior of memory operations in a multi-threaded context.
+Rather than use SC, however, most compilers implement a *weaker* memory model that allows for behaviors satisfying `r1 == r2 == 0`.
+
+Although this behavior is less intuitive, a weaker model than SC is necessary for two reasons:
+- hardware may reorder memory operations in such a way that contradicts the SC model.
+- thread-oblivious compiler optimizations don’t need to preserve SC. Thus, it is legal for memory operations reordering to break SC.
+
+
+## Pthreads’ Partial Memory Model
+Pthreads’ memory model is SC in the absence of data races.
+In the presence of data races, however, the memory model is **formally undefined**: any behavior is allowed.
+
+In other words, Pthreads allows any behavior for programs with data races. 
+The reason behind this design decision is explained as follows:
+>Formal definitions of the memory model were *rejected as unreadable by the vast majority of programmers*.
+>In addition, most of the formal work in *the literature has concentrated on the memory as provided by the hardware as opposed to the application programmer* through the compiler and runtime system.
+>It was believed that a simple statement intuitive to most programmers would be most effective
+
+Thus, since the above example contains data races, any compiler-chosen behavior satisfies the specification, including one for which `r1 == r2 == 0` holds.
+
+
+## Fighting Compilers for Well-Defined Behavior
+To facilitate the writing of race-free, well-defined Pthreads programs, Pthreads offers synchronization primitives such as the **memory barrier** and the **mutex**.
+Synchronization enables the containment of shared memory operations in programmer-defined **critical sections**, where memory operations are defined to be *mutually exclusive*.
+
+Since C++ is thread-oblivous, compilers are not guaranteed to respect the semantics of synchronization primitives.
+In particular, reads and writes to shared variables may be *moved out of critical sections*, introducing data races and, subsequently, undefined behavior.
+Thus, Pthreads relies on special support from compilers. 
+
+Pthreads-compatible compilers generally attempt to preserve synchronization semantics by restricting memory reorderings around synchronization primitives, such as `pthread_mutex_lock()`.
```

Comment:
Yep. I wanted to create suspense by saying that one *might think* that reads/writes would easily be moved out, but then have Pthreads-compilers save the day with opaque function treatment.

I now realize that opaque function treatment isn't really a strategy, but a convenient necessity, since Pthreads is an external library.